src/create/schema/index.ts

```diff
@@ -288,15 +288,18 @@ export const createSchemaOrRef = <
 
   const schemaRef = zodSchema._def.openapi?.ref ?? component?.ref;
 
+  let newState;
+
   if (zodSchema._def.openapi?.ref || component?.type === 'partial') {
     state.components.schemas.set(zodSchema, {
       type: 'inProgress',
       ref: (zodSchema._def.openapi?.ref ?? component?.ref) as string,
     });
+    newState = newSchemaState({ ...state, path: [], visited: new Set() });
```

Comment:
@samchungy when we encounter a partial ref'ed schema for the first time we need to discard the path and visited set and generate it as if it were the root.  Going back through the debugger to try to figure out how to explain what happens otherwise...

Reply:
Okay here's the error that happens with my updated test if we don't discard the visited set:

```
/user > get > responses > 200 > content > application/json > schema > property: posts > optional > array items > lazy schema > property: user_fields > optional > lazy schema > property: posts > optional needs to be registered because it's circularly referenced
```

Reply:
So my test has

```
    const UserSchema2: ZodType<User2> = BaseUser.extend({
      posts: PostArray.optional(),
      posts_fields: z.array(z.lazy(() => PartialPostSchema2)).optional(),
      comments: PostArray.optional(),
      comments_fields: z.array(z.lazy(() => PartialPostSchema2)).optional(),
    }); // no .openapi() for this test!
    const PartialUserSchema2: ZodType<Partial<User2>> = (
      UserSchema2 as any
    ).partial();
```

Because of the way `.partial()` works both `UserSchema2` and `PartialUserSchema2` are sharing the same `PostArray.optional()`.

So what happens is it starts generating `UserSchema2` and adds that `PostArray.optional()` to visited; then `UserSchema2 > posts > user_fields > Posts` returns to that same `PostArray.optional()` instance, thinking it's already been visited.

But `PartialUserSchema2` is registered as another referenced schema.  So we should be treating it as a new root generation, starting with an empty visited set.